File: src/client/plugins/optimizer/python/heuristics.py
import math
import random
import signal
from typing import Callable, cast

import numpy as np

from .calc import ConstraintDict, get_scale
from .solver import objective, pack
from .constraints import generate_constraints, select_initial_scale
from .problem import Circle, Hierarchy


def generate_candidate(target: int, hierarchies: list[Hierarchy]) -> list[np.ndarray]:
	vectors: list[np.ndarray] = []
	growth = target ** (1 / len(hierarchies))
	num = growth
	total = _estimate_total(target, growth, len(hierarchies))
	generated = 0

	def callback():
		nonlocal generated
		print(f'{{"event": "candidate", "data": [{generated}, {total}]}}')
		generated += 1

	context = GenerateContext(callback)
	handler = signal.signal(signal.SIGINT, context.handler)

	callback()
	last_hierarchy = None
	for hierarchy in hierarchies:
		context.set_hierarchy(hierarchy)
		if len(vectors) == 0:
			vectors = _generate_candidate(math.floor(num), context)
		elif last_hierarchy:
			vectors = _generate_next_level(vectors, context, last_hierarchy, num, target)
		last_hierarchy = hierarchy
		num *= growth

	vectors.sort(key=objective)
	signal.signal(signal.SIGINT, handler)
	return vectors[:target]


class GenerateContext:
	def __init__(self, callback: Callable):
		self.hierarchy: Hierarchy
		self.constraints: list[ConstraintDict] = []
		self.callback = callback
		self.interrupted = False

	def set_hierarchy(self, hierarchy: Hierarchy):
		self.hierarchy = hierarchy
		self.constraints = generate_constraints(hierarchy)

	def handler(self, _, __):
		self.interrupted = True


def _generate_next_level(
	vectors: list[np.ndarray], context: GenerateContext, last_hierarchy: Hierarchy, num: float, target: int
) -> list[np.ndarray]:
	next_vec: list[np.ndarray] = []
	num_per_vec = round(num / len(vectors))
	for vec in vectors:
		circles = _make_circles(vec, context.hierarchy, last_hierarchy)
		n = min(num_per_vec, target - len(next_vec))
		candidates = _generate_candidate(n, context, circles)
		next_vec.extend(candidates)
	return next_vec


def _estimate_total(target: int, growth: float, rounds: int) -> int:
	total = 0
	vectors = 0
	num = growth
	for i in range(rounds):
		if i == 0:
			vectors = math.floor(num)
		else:
			next_vec = 0
			num_per_vec = round(num / vectors)
			for _ in range(vectors):
				n = min(num_per_vec, target - next_vec)
				next_vec += n
			vectors = next_vec
		total += vectors
		num *= growth
	return total


def _make_circles(vec, hierarchy: Hierarchy, last_hierarchy: Hierarchy) -> list[Circle]:
	circles: list[Circle] = []
	for f in hierarchy.flaps:
		parent = hierarchy.parent_map.get(f.id, None)
		if parent is not None:
			n = next(i for i, x in enumerate(last_hierarchy.flaps) if x.id == parent.id)
			circles.append(Circle(vec[n * 2], vec[n * 2 + 1], parent.radius * vec[-1]))
		else:
			n = next(i for i, x in enumerate(last_hierarchy.flaps) if x.id == f.id)
			circles.append(Circle(vec[n * 2], vec[n * 2 + 1], 0))
	return circles


def _generate_candidate(target: int, context: GenerateContext, circles=None) -> list[np.ndarray]:
	vectors: list[np.ndarray] = []
	while len(vectors) < target:
		vec = _generate_random_candidate(context.hierarchy, circles)
		if not context.interrupted:
			result = pack(vec, context.constraints)
			if not result.success:
				continue
			vec = result.x
		vectors.append(vec)
		context.callback()
	return vectors


def _generate_random_candidate(hierarchy: Hierarchy, circles) -> np.ndarray:
	result = None
	for _ in range(4):
		if circles is None:
			vec: np.ndarray = np.random.rand(len(hierarchy.flaps) * 2 + 1)
		else:
			vec = np.array(_generate_in_circles(circles))
		select_initial_scale(vec, hierarchy)
		if result is None or get_scale(vec) < get_scale(result):
			result = vec
	return cast(np.ndarray, result)


def _generate_in_circles(circles: list[Circle]) -> list[float]:
	vec: list[float] = []
	for c in circles:
		if c.radius == 0:
			vec += [c.x, c.y]
		else:
			theta = random.uniform(0, 2 * math.pi)
			r = c.radius * math.sqrt(random.uniform(0, 1))  # uniform with respect to area
			vec += [c.x + r * math.cos(theta), c.y + r * math.sin(theta)]
	vec += [0]
	return vec

File: src/client/plugins/optimizer/src/math/heuristics.cpp

#include "heuristics.h"
#include "global/interrupt.h"
#include "scale.h"
#include "solver.h"

#include <algorithm>
#include <emscripten.h>
#include <iostream>

using namespace std;

using Vec = vector<double>;
using VecList = vector<Vec>;

struct Circle {
	double x;
	double y;
	double radius;
};

int find_index_by_id(const vector<Flap> &flaps, int id) {
	for (int i = 0; i < Shared::flap_count; i++) {
		if (flaps[i].id == id) return i;
	}
	throw std::runtime_error("Flap not found");
}

vector<Circle> make_circles(const Vec &vec, const Hierarchy &hierarchy, const Hierarchy &last_hierarchy) {
	vector<Circle> circles;
	circles.reserve(hierarchy.flaps.size());
	for (auto &f : hierarchy.flaps) {
		auto parent = hierarchy.parent_map.find(f.id);
		if (parent != hierarchy.parent_map.end()) {
			int n = find_index_by_id(last_hierarchy.flaps, parent->second->id);
			circles.emplace_back(vec[n * 2], vec[n * 2 + 1], parent->second->radius * vec.back());
		} else {
			int n = find_index_by_id(last_hierarchy.flaps, f.id);
			circles.emplace_back(vec[n * 2], vec[n * 2 + 1], 0);
		}
	}
	return circles;
}

Vec generate_in_circles(const vector<Circle> *circles) {
	Vec vec;
	vec.reserve(circles->size() * 2 + 1);
	for (auto &c : *circles) {
		if (c.radius == 0) {
			vec.push_back(c.x);
			vec.push_back(c.y);
		} else {
			auto theta = random01() * 2 * M_PI;
			auto r = c.radius * sqrt(random01()); // uniform with respect to area
			vec.push_back(c.x + r * cos(theta));
			vec.push_back(c.y + r * sin(theta));
		}
	}
	vec.push_back(0);
	return vec;
}

int estimate_total(int target, double growth, int rounds) {
	int total = 0;
	int vectors = 0;
	auto num = growth;
	for (int i = 0; i < rounds; i++) {
		if (i == 0) {
			vectors = floor(num);
		} else {
			int next_vec = 0;
			int num_per_vec = round(num / vectors);
			for (int j = 0; j < vectors; j++) {
				next_vec += min(num_per_vec, target - next_vec);
			}
			vectors = next_vec;
		}
		total += vectors;
		num *= growth;
	}
	return total;
}

class GenerateContext {
  public:
	GenerateContext(int total) : total(total) {}

	// We need the following because ConstraintList is holding unique_ptr
	GenerateContext(const GenerateContext &) = delete;
	GenerateContext &operator=(const GenerateContext &) = delete;

	void set_hierarchy(const Hierarchy *h) {
		hierarchy = h;
		Shared::setup(h->flaps.size()); // Next line depends on this
		constraints = h->generate_constraints(nullptr);
	}

	void callback() {
		cout << R"({"event": "candidate", "data": [)" << generated << ", " << total << "]}" << endl;
		generated++;
	}

	const Hierarchy *hierarchy;
	bool interrupted{false};
	ConstraintList constraints;

  private:
	int total;
	int generated{0};
};

Vec generate_random_vector(int size) {
	Vec result;
	result.reserve(size);
	for (int i = 0; i < size; i++) result.push_back(random01());
	return result;
}

Vec generate_random_candidate(const Hierarchy *hierarchy, const vector<Circle> *circles) {
	Vec result;
	for (int i = 0; i < 4; i++) {
		auto vec = circles ? generate_in_circles(circles) : generate_random_vector(hierarchy->flaps.size() * 2 + 1);
		setup_initial_scale(vec, hierarchy);
		if (result.empty() || get_scale(vec) < get_scale(result)) result = std::move(vec);
	}
	return result;
}

VecList generate_candidate(int target, GenerateContext &context, const vector<Circle> *circles) {
	VecList vectors;
	while (vectors.size() < target) {
		auto vec = generate_random_candidate(context.hierarchy, circles);
		if (!context.interrupted) {
			auto result = pack(vec, context.constraints);
			if (!result.success) continue;
			vec = result.x;
			if (check_interrupt()) context.interrupted = true;
		}
		vectors.push_back(vec);
		context.callback();
	}
	return vectors;
}

VecList generate_next_level(const VecList &vectors, GenerateContext &context, const Hierarchy &last_hierarchy, double num, int target) {
	VecList next_level;
	int num_per_vec = round(num / vectors.size());
	for (auto &vec : vectors) {
		auto circles = make_circles(vec, *context.hierarchy, last_hierarchy);
		auto n = min(num_per_vec, target - (int)next_level.size());
		auto candidates = generate_candidate(n, context, &circles);
		next_level.insert(next_level.end(), candidates.begin(), candidates.end());
	}
	return next_level;
}

VecList generate_candidate(int target, const vector<Hierarchy> &hierarchies) {
	VecList vectors;

	double growth = pow(target, 1.0 / hierarchies.size());
	double num = growth;
	int total = estimate_total(target, growth, hierarchies.size());
	int generated = 0;

	auto context = GenerateContext(total);
	context.callback();

	const Hierarchy *last_hierarchy = nullptr;
	for (auto &hierarchy : hierarchies) {
		context.set_hierarchy(&hierarchy);
		if (vectors.empty()) {
			vectors = generate_candidate(floor(num), context, nullptr);
		} else if (last_hierarchy) {
			vectors = generate_next_level(vectors, context, *last_hierarchy, num, target);
		}
		last_hierarchy = &hierarchy;
		num *= growth;
	}

	if (vectors.size() > target) {
		VecList sorted(target);
		partial_sort(vectors.begin(), vectors.begin() + target, vectors.end(), [](Vec a, Vec b) {
			return a[Shared::last] > b[Shared::last];
		});
		vectors.erase(vectors.begin() + target, vectors.end());
	}
	return vectors;
}
